### PR TITLE
 Access-Codes : add permissions, related endpoint docs and dredd tests

### LINKED ISSUE #4335: Add permissions to access codes API
Permissions missing from Access codes API.
Add related endpoint docs and dredd tests. [Parent issue #4009]

### PR DESCRIPTION
**Pull request by enigmaeth**
Fixes #4335.

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `nextgen` branch.
- [x] The unit tests pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

#### Changes proposed in this pull request:

- Add permissions according to dev handbook
- Add related endpoints docs
- Add related endpoints dredd tests

### CODE REVIEW

**Comment by enigmaeth** [comment]
@fossasia/open-event-server-devs The build hasn't started yet here. It passes on my repo. Here's the build link: https://travis-ci.org/enigmaeth/open-event-orga-server/builds/264478583. Please review :D

File: app/api/access_codes.py [hunk 1]
```diff
@@ -118,70 +140,56 @@ def query(self, view_kwargs):
...
+        Super-Admin/Admin can not delete an access code
+        Co-organizer can delete an access code
+        :param access_code:
+        :param view_kwargs:
+        :return:
+        """
+        if has_access('is_admin') and not has_access('is_coorganizer', event_id=view_kwargs['event_id']):
+            raise ForbiddenException({'source': ''}, 'You are not authorized')
 
     decorators = (jwt_required, )
```
Comment (shubham-padia) [inline app/api/access_codes.py]: Why not use `is_coorganizer` for respective methods from the permission manager itself into the decorators

Comment (enigmaeth) [inline app/api/access_codes.py]: @shubham-padia Resolved.

File: app/api/access_codes.py [hunk 2]
```diff
@@ -118,70 +141,56 @@ def query(self, view_kwargs):
...
     """
-    def before_get_object(self, view_kwargs):
-        if view_kwargs.get('ticket_id'):
-            ticket = safe_query(self, Ticket, 'id', view_kwargs['ticket_id'], 'ticket_id')
-            if ticket.access_code_id:
-                view_kwargs['id'] = ticket.access_code_id
-            else:
-                view_kwargs['id'] = None
+    @jwt_required
+    def before_get(self, args, kwargs):
```
Comment (shubham-padia) [inline app/api/access_codes.py]: no need for this

Comment (enigmaeth) [inline app/api/access_codes.py]: Only registered user can view the access-code details, so :sweat_smile:

File: app/api/access_codes.py [hunk 3]
```diff
@@ -118,70 +141,56 @@ def query(self, view_kwargs):
...
+        """
+        Method to check for admin access before deleting an access code
+        Super-Admin/Admin can not delete an access code
+        Co-organizer can delete an access code
+        :param access_code:
+        :param view_kwargs:
+        :return:
+        """
+        if has_access('is_admin'):
+            raise ForbiddenException({'source': ''}, 'You are not authorized')
```
Comment (shubham-padia) [inline app/api/access_codes.py]: Add a new function in permission manager called maybe `is_coorganizer_not_accessible_by_admin` or any other better name you can come up with. This user should check if the user is a co-organizer or organizer only and should not grant permission even if the user is admin. Then remove the `before_delete_object` method and instead use permission manager in decorators specifying the method delete.

File: app/api/access_codes.py (hunk 2 above)
Comment (shubham-padia) [inline app/api/access_codes.py]: @enigmaeth If the user is coorganizer level, the access-code should be accessible by id, otherwise they should not be accessible by id, as quoted in the handbook.
> Only valid and non-exhausted ones and accessed via the access code itself (not ID).

please cover the above point in a separate follow-up issue, for now remove the above and use `is_coorganizer` permission for get endpooint also.

File: app/api/access_codes.py [hunk 4]
```diff
@@ -118,70 +141,56 @@ def query(self, view_kwargs):
...
-        data['user_id'] = current_identity.id
-
-    def after_create_object(self, obj, data, view_kwargs):
-        if view_kwargs.get('ticket_id'):
-            ticket = safe_query(self, Ticket, 'id', view_kwargs['ticket_id'], 'ticket_id')
-            ticket.access_codes.append(obj)
-            self.session.commit()
-
     view_kwargs = True
     decorators = (jwt_required, )
```
Comment (shubham-padia) [inline app/api/access_codes.py]: `is_coorganizer` here also I think, please correct me if wrong :)

**Comment by enigmaeth** [comment]
@shubham-padia Issues resolved. Please review :)

File: app/api/access_codes.py [hunk 5]
```diff
@@ -1,27 +1,50 @@
...
+        before post method to check for required relationships and permissions
+        :param args:
+        :param kwargs:
+        :param data:
+        :return:
+        """
+        require_relationship(['event', 'user'], data)
+        if not has_access('is_coorganizer', event_id=data['event']):
+            raise ObjectNotFound({'parameter': 'event_id'},
+                                 "Event: {} not found".format(data['event_id']))
```
Comment (shubham-padia) [inline app/api/access_codes.py]: The above will be `.format(data['event'])`

File: app/api/access_codes.py [hunk 6]
```diff
@@ -36,70 +59,37 @@ def query(self, view_kwargs):
...
-        if view_kwargs.get('ticket_id'):
-            ticket = safe_query(self, Ticket, 'id', view_kwargs['ticket_id'], 'ticket_id')
-            ticket.access_codes.append(obj)
-            self.session.commit()
-
     view_kwargs = True
-    decorators = (jwt_required, )
+    decorators = (api.has_permission('is_coorganizer', fetch='event_id',
+                  fetch_as="event_id", model=AccessCode,
+                  leave_if=lambda a: a.get('user_id') or a.get('ticket_id')), )
```
Comment (shubham-padia) [inline app/api/access_codes.py]: Add relevant permissions in the query object itself in case user_id or ticket_id is there.

Comment (enigmaeth) [inline app/api/access_codes.py]: Not sure if this the right way to filter only access-codes for which co-orga access is present. (event_id is required relationship, so it will always be there for an access-code). Please check.

Comment (shubham-padia) [inline app/api/access_codes.py]: Can you rephrase your statement a bit please, cannot get it :sweat_smile:

Comment (enigmaeth) [inline app/api/access_codes.py]: I mean that for all the three endpoints : `events/id/access-code`, `tickets/id/access-code`, `users/id/access-code`, we need to return only those access-codes which belong to an event of which the current user is co-organizer. So is `query_ = query_.filter(has_access('is_coorganizer', event_id=AccessCode.event_id))` a right way of doing that? If not, please suggest what else should be done :sweat_smile:!

Comment (shubham-padia) [inline app/api/access_codes.py]: See the changes addition of permissions in attendees API in #4357 for reference.

Comment (enigmaeth) [inline app/api/access_codes.py]: @shubham-padia Resolved.
